subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/CrossVersionTestEngine.java

```diff
@@ -0,0 +1,234 @@
+import org.gradle.util.GradleVersion;
+import org.junit.platform.engine.ConfigurationParameters;
+import org.junit.platform.engine.DiscoveryFilter;
+import org.junit.platform.engine.DiscoverySelector;
+import org.junit.platform.engine.EngineDiscoveryListener;
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.ExecutionRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.discovery.ClassSelector;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.support.descriptor.EngineDescriptor;
+import org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine;
+import org.spockframework.runtime.RunContext;
+import org.spockframework.runtime.SpockEngine;
+import org.spockframework.runtime.SpockEngineDescriptor;
+import org.spockframework.runtime.SpockExecutionContext;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.gradle.integtests.fixtures.compatibility.AbstractContextualMultiVersionTestInterceptor.VERSIONS_SYSPROP_NAME;
+
+
+public class CrossVersionTestEngine extends HierarchicalTestEngine<SpockExecutionContext> {
+
+    private final TestEngine delegateEngine = new SpockEngine();
+    private final List<TestVariant> variants = new ArrayList<TestVariant>();
+
+    @Override
+    public String getId() {
+        return "cross-version-test-engine";
+    }
+
+    @Override
+    public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
+        String crossVersionTest = System.getProperty("org.gradle.integtest.crossVersion");
+        if (crossVersionTest == null) {
+            return new EngineDescriptor(uniqueId, "skip");
+        }
+
+        buildTestVariants(uniqueId, discoveryRequest);
+        return discoverTests(uniqueId);
+    }
+
+    @Override
+    protected SpockExecutionContext createExecutionContext(ExecutionRequest request) {
+        return new SpockExecutionContext(request.getEngineExecutionListener());
+    }
+
+    private void buildTestVariants(UniqueId rootId, EngineDiscoveryRequest discoveryRequest) {
+        variants.add(TestVariant.tapiCurrent(rootId, discoveryRequest));
+        variants.add(TestVariant.selected(rootId, discoveryRequest));
+        variants.add(TestVariant.tapiTargetLoaded(rootId, discoveryRequest));
+    }
+
+    private TestDescriptor discoverTests(UniqueId uniqueId) {
+        EngineDescriptor rootDescriptor = new SpockEngineDescriptor(uniqueId, RunContext.get());
+        for (TestVariant testVariant : variants) {
+            for (TestDescriptor test : testVariant.discover(delegateEngine).getChildren()) {
+                rootDescriptor.addChild(test);
+            }
+        }
+        return rootDescriptor;
```

Comment:
Have you tried instead creating your own engine descriptor and adding the descriptors returned by Spock as children (maybe decorated to adjust the display name) to that? That way, we would depend on fewer internal classes of Spock, i.e. `SpockEngineDescriptor` and `RunContext`.

Reply:
Yes - I started that way and it kinda worked execution-wise, but the test aggregation then became confusing, like
![image](https://user-images.githubusercontent.com/5560673/105042229-a8ce6b00-5a6c-11eb-8123-4b219e1e0e88.png)
It looks as if each new thread of test execution then creates a new subtree with root named `Spock`. Whereas with current implementation, the test methods are aggregated under the class name they belong to, even if they come from different "variants".

Reply:
Would it help if you decorated the `TestDescriptor` so instead of "Spock" it would say whatever is relevant for the variant?

Reply:
That is also tricky - the first thing Spock engine does is a cast `(SpockEngineDescriptor)request.getRootTestDescriptor()`. And I can not decorate `SpockEngineDescriptor` to override the display name because:
1. `getDisplayName()` is final in `AbstractTestDescriptor`
2. `SpockEngineDescriptor` sets the display name to "Spock" in its constructor - can't override it there either.

What I can do instead and what seems to partially work is implement a new TestDescriptor that delegates everything but `getDisplayName()` to the Spock engine descriptor and then in my test engine, extract the original Spock descriptor like this:
```
for (TestDescriptor delegateDescriptor : request.getRootTestDescriptor().getChildren()) {
            delegateEngine.execute(new ExecutionRequest(((DelegatingTestDescriptor)delegateDescriptor).getDelegate(), request.getEngineExecutionListener(), request.getConfigurationParameters()));
        }
```

What I still do not like with this approach is that the tests still seem to be aggregated per thread of execution:
![image](https://user-images.githubusercontent.com/5560673/105169968-f27a8c80-5b24-11eb-9b05-9129627f80b1.png)
And here's the resulting [build scan](https://ge.gradle.org/s/opxdw3ltmglca/tests?toggled=WyI6dG9vbGluZy1hcGktZW1iZWRkZWRjcm9zc3ZlcnNpb250ZXN0LW9yZy5ncmFkbGUuaW50ZWd0ZXN0cy50b29saW5nLnI2NSJd) - looks better there, despite the variant containers reported as unknown classes separately.

Reply:
I see, but could you store the original `SpockEngineDescriptor` in your own `AbstractTestDescriptor` and create a new `ExecutionRequest` from it?